stats: prevent unused counters from leaking across hot restart

fredlas opened the pull request:
During hot restart, stores counters imported from the parent in a temporary scope. When hot restart concludes, that scope is dropped. The result is that any counter the child accessed (independently of the hot restart merging) will be preserved by the separate reference that access created, but any counter not so accessed will be dropped. This brings the hot restart counter behavior back to how it was before shared memory was removed in #5910.

Risk Level: low
Testing: updated test/common/stats/stat_merger_test.cc.

More fully fixes #4974.

fredlas commented:
/assign jmarantz
/assign mattklein123

mattklein123 commented on include/envoy/stats/store.h:
```diff
@@ -43,6 +43,11 @@ class Store : public Scope {
+  /**
+   * @return whether any known counter exists with this name.
+   */
+  virtual bool counterExists(const std::string& counter_name) PURE;
```
cc @ahedberg for intersection with https://github.com/envoyproxy/envoy/pull/6712. Can we intersect with the APIs being added there so we don't duplicate?

fredlas replied:
This one is test only, so it will definitely be best to just get rid of counterExists, and use those ones in here. That will be a very easy switch; left a todo.

mattklein123 commented on source/common/stats/stat_merger.h:
```diff
@@ -36,7 +36,26 @@ class StatMerger {
   StatNameHashMap<uint64_t> parent_gauge_values_;
-  Stats::Store& target_store_;
+  Store& target_store_;
+  // A stats Scope for our in-the-merging-process counters to live in. Scopes conceptually hold
```
We need to do the same for gauges, right? To avoid gauge leaks?

fredlas replied:
Done.

mattklein123 commented on tools/spelling_dictionary.txt:
```diff
@@ -236,6 +236,8 @@ SRDS
 SRV
 SS
 SSL
+STATRICIA
```
nit: can we pick a different dictionary word to avoid adding this to the dictionary?

fredlas replied:
Done.

fredlas commented:
Ran all tests this time. All the CI should be fixed.

mattklein123 commented:
@fredlas ASAN failure looks legit.

ahedberg commented:
> Thanks, this looks great, assuming @ahedberg is OK with this merging first? @jmarantz can you take a quick pass?

Yup, SGTM. Fred's TODO SGTM too.

fredlas commented:
Fixed. I ran into the same thing a while back, and kind of blindly got around it. Looking closer this time, it looks like ThreadLocalStoreImpl has two bools shutting_down_ and merge_in_progress_, and there ought to be one for threading_ever_started_. So really, it should probably just be an enum. I will go make that change, unless there is a reason not to that immediately comes to mind.

mattklein123 commented:
@fredlas please fix if you can clean up. We can just do it in this PR?

/wait-any

fredlas commented:
I think it deserves its own PR. It will be a somewhat nontrivial change in ThreadLocalStoreImpl. I know this PR already adds some stuff in that file, but 1) they're just self-contained little checks, and 2) they're going to be gone very soon anyways.

fredlas commented:
Haha sorry, changed my mind. Turns out the state transitions don't 100% make sense to switch to an enum. Just adding the 3rd bool is super simple, so I have now added that to this PR.

mattklein123 commented:
/retest

mattklein123 commented on source/common/stats/stat_merger.h:
```diff
@@ -36,7 +36,25 @@ class StatMerger {
+  // preserve all stats throughout the hot restart. Then, when the restart completes, dropping
+  // the scope will drop exactly those stats whose names have not already been accessed through
+  // another store/scope.
+  ScopePtr temp_scope_;
```
It makes me happy how easy this was to accomplish. Pretty slick!

fredlas replied:
Yeah, it certainly is. Thanks for thinking of this approach!

jmarantz commented on test/common/stats/stat_merger_test.cc:
```diff
@@ -164,6 +154,40 @@ TEST_F(StatMergerTest, exclusionsNotImported) {
+  // We accessed 0 and 1 above, but not 2. Now that StatMerger has been destroyed,
+  // 2 should be gone.
+  // TODO(fredlas) replace counterExists with Scope::getCounter once #6712 is merged.
+  EXPECT_TRUE(store.counterExists("newcounter0"));
```
it looks like the only call-sites are on ThreadLocalStore and not on any other subclasses of Store. Move the temp helper method out of the interface and have it just in that specific class?

also ok in follow-up.

fredlas replied:
(obsolete)

mattklein123 commented:
Fine for me to wait for @jmarantz comments. They all seem like great things to do.

/wait

fredlas commented:
All of that code is going to be deleted when #6712 merges. I don't think it's worth perfecting. If #6712 is going to take a while, maybe splitting out just the `getCounter` part of it to be submitted separately would be the way to go.

mattklein123 commented:
> All of that code is going to be deleted when #6712 merges. I don't think it's worth perfecting. If #6712 is going to take a while, maybe splitting out just the getCounter part of it to be submitted separately would be the way to go.

I agree, but why add it in the first place given that we don't know when the other PR will merge? Per @jmarantz we can just have a method directly on the ThreadLocalStore that does what we need and is marked for testing?

jmarantz commented:
Moreover it occurred to me after reflecting on the code that these new functions aren't needed, as there is already TestUtility::findCounter() and findGauge, declared in test/test_common/utility.h

fredlas commented on source/common/stats/isolated_store_impl.h:
```diff
@@ -49,6 +49,8 @@ template <class Base> class IsolatedStatsCache {
     return vec;
   }
 
+  bool statExists(StatName name) { return stats_.find(name) != stats_.end(); }
```
(obsolete)

fredlas commented on source/common/stats/thread_local_store.cc:
```diff
@@ -146,8 +146,33 @@ std::vector<ParentHistogramSharedPtr> ThreadLocalStoreImpl::histograms() const {
+  Thread::LockGuard lock(lock_);
+  for (ScopeImpl* scope : scopes_) {
+    if (scope->central_cache_.counters_.find(
+            StatNameManagedStorage(counter_name, symbolTable()).statName()) !=
```
(obsolete)

fredlas commented on source/common/stats/thread_local_store.cc:
```diff
@@ -146,8 +146,33 @@ std::vector<ParentHistogramSharedPtr> ThreadLocalStoreImpl::histograms() const {
+  Thread::LockGuard lock(lock_);
+  for (ScopeImpl* scope : scopes_) {
+    if (scope->central_cache_.gauges_.find(
+            StatNameManagedStorage(gauge_name, symbolTable()).statName()) !=
```
(obsolete)

fredlas commented:
Nice find, thanks! Switched to those.

jmarantz commented on test/common/stats/stat_merger_test.cc:
```diff
@@ -164,6 +151,39 @@ TEST_F(StatMergerTest, exclusionsNotImported) {
+  }
+  // We accessed 0 and 1 above, but not 2. Now that StatMerger has been destroyed,
+  // 2 should be gone.
+  EXPECT_TRUE(TestUtility::findCounter(store, "newcounter0"));
```
I'd probably write it as `EXPECT_TRUE(TestUtility::findCounter(store, "newcounter0") != nullptr)` or maybe `EXPECT_NE(TestUtility::findCounter(store, "newcounter0"), nullptr)` if that compiles, but meh :)

rgs1 commented:
@fredlas @jmarantz fwiw, we picked up this change but we are still seeing metrics drop over a hot restart. I'll dig into it a bit more, but wanted to give you an early heads up.

fredlas commented:
By "drop", do you mean values going to 0, or metrics going from existing to not existing? Either version would be different from the issue solved in this PR, so this discussion should probably move to an issue, but it does sound like something to discuss.

rgs1 commented:
This is how it looks for `cluster.${cluster}.health_check.failure` during a hot restart:

![image](https://user-images.githubusercontent.com/1349891/57641888-1af66900-7584-11e9-88ff-2b4ae8dcb721.png)

I am wondering if the move from shared mem to copying stats over IPC would allow for such a drop & recovery to happen. Given that copying stats over takes time... 

cc: @fishcakez

fredlas commented:
That problem is not related to this PR; we should discuss it elsewhere.

mattklein123 commented:
@rgs1 I have an idea of why this is happening but per @fredlas it's unrelated to this PR. Can you open a fresh issue and we can discuss there?

rgs1 commented:
@mattklein123 yeah lets chat here: https://github.com/envoyproxy/envoy/issues/6924

@fredlas thanks!
